pkg/storegateway/indexcache/cache.go

```diff
@@ -76,6 +81,33 @@ type IndexCache interface {
 	FetchLabelValues(ctx context.Context, userID string, blockID ulid.ULID, labelName string, matchersKey LabelMatchersKey) ([]byte, bool)
 }
 
+// PostingsKey represents a canonical key for a []storage.SeriesRef slice
+type PostingsKey string
+
+// CanonicalPostingsKey creates a canonical version of PostingsKey
+func CanonicalPostingsKey(postings []storage.SeriesRef) PostingsKey {
```

Comment:
Can we have a benchmark to see how long it take with a different number of postings (e.g. from 10 to 1M with x10 steps)?

Reply:
I added `BenchmarkCanonicalPostingsKey` in a0366f5aad1a1fa548f83377ed2cf34bea2d8ffa

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway/indexcache
BenchmarkCanonicalPostingsKey
BenchmarkCanonicalPostingsKey/10_postings
BenchmarkCanonicalPostingsKey/10_postings-10         	  394296	      2971 ns/op
BenchmarkCanonicalPostingsKey/100_postings
BenchmarkCanonicalPostingsKey/100_postings-10        	   46702	     25242 ns/op
BenchmarkCanonicalPostingsKey/1000_postings
BenchmarkCanonicalPostingsKey/1000_postings-10       	   12168	     98330 ns/op
BenchmarkCanonicalPostingsKey/10000_postings
BenchmarkCanonicalPostingsKey/10000_postings-10      	    1332	    895424 ns/op
BenchmarkCanonicalPostingsKey/100000_postings
BenchmarkCanonicalPostingsKey/100000_postings-10     	     135	   8807268 ns/op
BenchmarkCanonicalPostingsKey/1000000_postings
BenchmarkCanonicalPostingsKey/1000000_postings-10    	      13	  87064817 ns/op
PASS
```

for reference on what will take time to hash matchers vs postings - this is `BenchmarkCanonicalLabelMatchersKey`

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway/indexcache
BenchmarkCanonicalLabelMatchersKey
BenchmarkCanonicalLabelMatchersKey/1_matchers
BenchmarkCanonicalLabelMatchersKey/1_matchers-10         	13383583	        88.54 ns/op
BenchmarkCanonicalLabelMatchersKey/5_matchers
BenchmarkCanonicalLabelMatchersKey/5_matchers-10         	 4018947	       297.2 ns/op
BenchmarkCanonicalLabelMatchersKey/10_matchers
BenchmarkCanonicalLabelMatchersKey/10_matchers-10        	 1568452	       744.9 ns/op
BenchmarkCanonicalLabelMatchersKey/20_matchers
BenchmarkCanonicalLabelMatchersKey/20_matchers-10        	  737541	      1671 ns/op
PASS
```

it appears that sorting is 72% of the work - this is a profile of the whole benchmark (all 6 cases)
![Screenshot 2022-12-13 at 20 53 17](https://user-images.githubusercontent.com/21020035/207430899-c7fdfacd-f916-48f1-9e49-4e838d1dba73.png)

in a test cluster using almost exclusively the cache to serve requests calculating the postings key is 0.7% of CPU

![Screenshot 2022-12-13 at 20 55 15](https://user-images.githubusercontent.com/21020035/207431257-f0900d9f-1f33-46f0-92db-5329ed6cc0fc.png)

this seems fairly minor to me 🤷 WDYT?